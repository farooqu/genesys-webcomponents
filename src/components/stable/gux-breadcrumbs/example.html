<h1>gux-breadcrumbs</h1>

<h2>Primary</h2>

<gux-breadcrumbs accent="primary">
  <gux-breadcrumb-item href="#">Department</gux-breadcrumb-item>
  <gux-breadcrumb-item>Category</gux-breadcrumb-item>
  <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
  <gux-breadcrumb-item>Current</gux-breadcrumb-item>
</gux-breadcrumbs>

<h2>Secondary</h2>

<gux-breadcrumbs accent="secondary">
  <gux-breadcrumb-item href="#">Department</gux-breadcrumb-item>
  <gux-breadcrumb-item>Category</gux-breadcrumb-item>
  <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
  <gux-breadcrumb-item>Current</gux-breadcrumb-item>
</gux-breadcrumbs>
