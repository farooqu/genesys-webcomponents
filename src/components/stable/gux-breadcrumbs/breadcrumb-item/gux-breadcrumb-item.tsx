import { Component, Element, h, JSX, Prop } from '@stencil/core';

import { logError } from '../../../../utils/error/log-error';

import { GuxBreadcrumbAccent } from '../gux-breadcrumbs.types';

@Component({
  styleUrl: 'gux-breadcrumb-item.less',
  tag: 'gux-breadcrumb-item',
  shadow: true
})
export class GuxBreadcrumbItem {
  @Element()
  private root: HTMLElement;

  @Prop()
  href: string;

  private getAccent(): GuxBreadcrumbAccent {
    const container = this.root.closest('gux-breadcrumbs');

    if (container) {
      return container.accent;
    } else {
      logError(
        'gux-breadcrumb-item',
        'This component must be a child of a gux-breadcrumbs component.'
      );
    }
  }

  private isLastBreadcrumb(): boolean {
    const parentNode = this.root.parentNode;
    const children = parentNode.children;

    return children[children.length - 1] === this.root;
  }

  private getBreadcrumb(): JSX.Element {
    if (!this.href || this.isLastBreadcrumb()) {
      return (
        <span class="gux-breadcrumb-content">
          <slot />
        </span>
      ) as JSX.Element;
    }

    return (
      <a class="gux-breadcrumb-content gux-link" href={this.href}>
        <slot />
      </a>
    ) as JSX.Element;
  }

  private getSeparatorIcon(accent: GuxBreadcrumbAccent): JSX.Element {
    if (this.isLastBreadcrumb()) {
      return null;
    }

    switch (accent) {
      case 'primary':
        return (
          <span class="gux-breadcrumb-separator" aria-hidden="true">
            /
          </span>
        ) as JSX.Element;
      case 'secondary':
        return (
          <gux-icon
            class="gux-breadcrumb-separator"
            icon-name="chevron-small-right"
            decorative
          ></gux-icon>
        ) as JSX.Element;
      default:
        return (
          <span class="gux-breadcrumb-separator" aria-hidden="true">
            /
          </span>
        ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    const accent: GuxBreadcrumbAccent = this.getAccent();

    return (
      <span class={`gux-breadcrumb-generation gux-${accent}`}>
        {this.getBreadcrumb()}
        {this.getSeparatorIcon(accent)}
      </span>
    ) as JSX.Element;
  }
}
