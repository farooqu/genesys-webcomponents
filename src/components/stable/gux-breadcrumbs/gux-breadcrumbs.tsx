import { Component, Element, h, JSX, Prop } from '@stencil/core';

import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import { trackComponent } from '../../../usage-tracking';

import breadcrumbsResources from './i18n/en.json';
import { GuxBreadcrumbAccent } from './gux-breadcrumbs.types';

@Component({
  styleUrl: 'gux-breadcrumbs.less',
  tag: 'gux-breadcrumbs',
  shadow: true
})
export class GuxBreadcrumbs {
  private i18n: GetI18nValue;

  @Element()
  private root: HTMLElement;

  @Prop()
  accent: GuxBreadcrumbAccent = 'primary';

  componentWillLoad(): void {
    trackComponent(this.root, { variant: this.accent });
  }

  async componentWillRender(): Promise<void> {
    this.i18n = await buildI18nForComponent(this.root, breadcrumbsResources);
  }

  render(): JSX.Element {
    return (
      <nav
        aria-label={this.i18n('breadcrumbs')}
        class="gux-breadcrumbs-container"
      >
        <slot />
      </nav>
    ) as JSX.Element;
  }
}
