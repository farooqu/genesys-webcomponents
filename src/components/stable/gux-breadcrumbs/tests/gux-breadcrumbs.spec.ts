jest.mock('../../../../utils/error/log-error', () => ({
  __esModule: true,
  logError: jest.fn()
}));

import { newSpecPage } from '@stencil/core/testing';

import { GuxBreadcrumbs } from '../gux-breadcrumbs';
import { GuxBreadcrumbItem } from '../breadcrumb-item/gux-breadcrumb-item';

import { logError } from '../../../../utils/error/log-error';

const components = [GuxBreadcrumbs, GuxBreadcrumbItem];
const language = 'en';

describe('gux-breadcrumbs', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should build', async () => {
    const page = await newSpecPage({
      components,
      html: `<gux-breadcrumbs></gux-breadcrumbs>`,
      language
    });

    expect(page.rootInstance).toBeInstanceOf(GuxBreadcrumbs);
  });

  describe('#render', () => {
    [
      `<gux-breadcrumbs></gux-breadcrumbs>`,
      `
        <gux-breadcrumbs>
          <gux-breadcrumb-item>Category</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
          <gux-breadcrumb-item>Current</gux-breadcrumb-item>
        </gux-breadcrumbs>
      `,
      `
        <gux-breadcrumbs>
          <gux-breadcrumb-item>Category</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Current</gux-breadcrumb-item>
        </gux-breadcrumbs>
      `,
      `
        <gux-breadcrumbs accent="primary">
          <gux-breadcrumb-item>Category</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
          <gux-breadcrumb-item>Current</gux-breadcrumb-item>
        </gux-breadcrumbs>
      `,
      `
        <gux-breadcrumbs accent="secondary">
          <gux-breadcrumb-item>Category</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
          <gux-breadcrumb-item>Current</gux-breadcrumb-item>
        </gux-breadcrumbs>
      `,
      `
        <gux-breadcrumbs accent="invalid">
          <gux-breadcrumb-item>Category</gux-breadcrumb-item>
          <gux-breadcrumb-item href="#">Parent</gux-breadcrumb-item>
          <gux-breadcrumb-item>Current</gux-breadcrumb-item>
        </gux-breadcrumbs>
      `
    ].forEach((html, index) => {
      it(`should render component as expected (${index + 1})`, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });
  });

  describe('valid slotting', () => {
    it('should not log an error if a gux-breadcrumb-item component is slotted in a gux-breadcrumb component', async () => {
      await newSpecPage({
        components,
        html: `
          <gux-breadcrumbs>
            <gux-breadcrumb-item>Grandparent</gux-breadcrumb-item>
            <gux-breadcrumb-item>Parent</gux-breadcrumb-item>
            <gux-breadcrumb-item>Child</gux-breadcrumb-item>
          </gux-breadcrumbs>
        `,
        language
      });

      expect(logError).not.toHaveBeenCalled();
    });

    it('should log an error if a gux-breadcrumb-item component is slotted in an element other than a gux-breadcrumb component', async () => {
      await newSpecPage({
        components,
        html: `
          <div>
            <gux-breadcrumb-item>Grandparent</gux-breadcrumb-item>
            <gux-breadcrumb-item>Parent</gux-breadcrumb-item>
            <gux-breadcrumb-item>Child</gux-breadcrumb-item>
          </div>
        `,
        language
      });

      expect(logError).toHaveBeenCalledTimes(3);
    });
  });
});
