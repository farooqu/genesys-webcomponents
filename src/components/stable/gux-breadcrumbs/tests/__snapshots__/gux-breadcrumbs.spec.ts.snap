// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-breadcrumbs #render should render component as expected (1) 1`] = `
<gux-breadcrumbs>
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container"></nav>
</gux-breadcrumbs>
`;

exports[`gux-breadcrumbs #render should render component as expected (2) 1`] = `
<gux-breadcrumbs>
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container">
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Category
        </span>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <a class="gux-breadcrumb-content gux-link" href="#">
          Parent
        </a>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Current
        </span>
      </span>
    </gux-breadcrumb-item>
  </nav>
</gux-breadcrumbs>
`;

exports[`gux-breadcrumbs #render should render component as expected (3) 1`] = `
<gux-breadcrumbs>
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container">
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Category
        </span>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <a class="gux-breadcrumb-content gux-link" href="#">
          Parent
        </a>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Current
        </span>
      </span>
    </gux-breadcrumb-item>
  </nav>
</gux-breadcrumbs>
`;

exports[`gux-breadcrumbs #render should render component as expected (4) 1`] = `
<gux-breadcrumbs accent="primary">
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container">
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Category
        </span>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <a class="gux-breadcrumb-content gux-link" href="#">
          Parent
        </a>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-primary">
        <span class="gux-breadcrumb-content">
          Current
        </span>
      </span>
    </gux-breadcrumb-item>
  </nav>
</gux-breadcrumbs>
`;

exports[`gux-breadcrumbs #render should render component as expected (5) 1`] = `
<gux-breadcrumbs accent="secondary">
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container">
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-secondary">
        <span class="gux-breadcrumb-content">
          Category
        </span>
        <gux-icon class="gux-breadcrumb-separator" decorative="" icon-name="chevron-sm-right"></gux-icon>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-secondary">
        <a class="gux-breadcrumb-content gux-link" href="#">
          Parent
        </a>
        <gux-icon class="gux-breadcrumb-separator" decorative="" icon-name="chevron-sm-right"></gux-icon>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-secondary">
        <span class="gux-breadcrumb-content">
          Current
        </span>
      </span>
    </gux-breadcrumb-item>
  </nav>
</gux-breadcrumbs>
`;

exports[`gux-breadcrumbs #render should render component as expected (6) 1`] = `
<gux-breadcrumbs accent="invalid">
  <!---->
  <nav aria-label="Breadcrumbs" class="gux-breadcrumbs-container">
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-invalid">
        <span class="gux-breadcrumb-content">
          Category
        </span>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item href="#">
      <!---->
      <span class="gux-breadcrumb-generation gux-invalid">
        <a class="gux-breadcrumb-content gux-link" href="#">
          Parent
        </a>
        <span aria-hidden="true" class="gux-breadcrumb-separator">
          /
        </span>
      </span>
    </gux-breadcrumb-item>
    <gux-breadcrumb-item>
      <!---->
      <span class="gux-breadcrumb-generation gux-invalid">
        <span class="gux-breadcrumb-content">
          Current
        </span>
      </span>
    </gux-breadcrumb-item>
  </nav>
</gux-breadcrumbs>
`;
